### ISSUE TITLE
Support Multiple Models

### ISSUE THREAD

**Issue by aldrinc** [vllm-project/vllm#299]
- Allow user to specify multiple models to download when loading server
- Allow user to switch between models 
- Allow user to load multiple models on the cluster (nice to have)

**Comment by zhuohan123** [comment]
For first and second feature request, why don't you just kill the old server and start a new one with the new model?

**Comment by aldrinc** [comment]
It’s what we are doing now but takes long time to download and load large models (33/65b). For train to deploy pipeline not possible to have zero downtime without multiple servers and blue green deployment strategy.

**Comment by zhuohan123** [comment]
The downloading should only happen at the very first time when using a model. However, the loading cost is unavoidable. Are you looking for something that can swap the model with zero downtime?

**Comment by gesanqiu** [comment]
I think multi-models is important to some bussiness logic like ensemble model, do you have any ideas I can reference, then I will try to implement it on vLLM.

**Comment by ft-algo** [comment]
Is there any progress on this feature?

**Comment by shixianc** [comment]
+1

**Comment by Ki6an** [comment]
+1

**Comment by corticalstack** [comment]
+1 For enterprise, instead of one monolithic, API-based LLM like GPT4, their strategy may be a collection of SLMs dedicated/fine-tuned to specific tasks. THis is why they will want to serve multiple models simultaneously, for e.g. phi2/mistral-7B/yi-34B. Could we have an update on this feature request please?

**Comment by chanchimin** [comment]
+1
how is the progress of this feature

**Comment by lenartgolob** [comment]
+1

**Comment by Shamepoo** [comment]
+1

**Comment by mjtechguy** [comment]
+1

**Comment by jvlinsta** [comment]
+1

**Comment by mohittalele** [comment]
+1

**Comment by ptrmayer** [comment]
+1

**Comment by tarson96** [comment]
++++1

**Comment by ptrmayer** [comment]
+1

**Comment by Luffyzm3D2Y** [comment]
+1

**Comment by srzer** [comment]
+1 desperately need multiple models

**Comment by gillan-rbih** [comment]
+1

**Comment by lizhipengpeng** [comment]
+1

**Comment by amitm02** [comment]
+1

**Comment by naturomics** [comment]
+1

**Comment by servient-ashwin** [comment]
What is the general thought process or strategy into implementing like this if this is open for contributions? Does the vLLM team or anyone have a roadmap or idea into it's implementation that someone can pick up?

This would be highly useful. All I can quickly think of is how tensorflow allows to specify memory used something similar to `gpu util` here on vLLM or using subprocesses within python (although I must admit I do not understand much about this topic in general anyway)

**Comment by zr-idol** [comment]
+1

**Comment by Sk4467** [comment]
One work around that I notice vLLM already gives is using Docker containers.
Use docker compose to serve multiple models that way two or more models can be served at a time based on your memory capabilities.

**Comment by ws-chenc** [comment]
+++1

**Comment by chen-j-ing** [comment]
+1+1

**Comment by ArtificialEU** [comment]
So i can have a complex vllm setup with batching and all but can't change the current model except spin a new container ? 😅

**Comment by pingbowen23** [comment]
+1

**Comment by JoursBleu** [comment]
+1

**Comment by nathan-weinberg** [comment]
+1

**Comment by kglmcodes** [comment]
+1

**Comment by youkaichao** [comment]
vLLM does not support multiple models.

for anyone who is interested, please pin up multiple vllm instances, and use routers like litellm https://docs.litellm.ai/docs/simple_proxy#load-balancing---multiple-instances-of-1-model to route among vllm instances.

**Comment by Uri-meir** [comment]
+1

**Comment by jdmuntacb** [comment]
+1 Good to have this.

**Comment by jbohnslav** [comment]
+1, at the very least would be great to see an example. Ray serve's vLLM example does not work with multiple models and tensor parallelism

**Comment by DA21S321D** [comment]
+1, I just need two models load separately

**Comment by alexzender** [comment]
+1

**Comment by seungduk-yanolja** [comment]
Would it be possible to switch models by utilizing RAM?
E.g.,
1. Load all models to RAM and load one of them from RAM to the GPU memory.
2. If the requested model is not the one loaded in the GPU memory, load it from the RAM.

I am not sure if it is possible.

**Comment by ashtonomy** [comment]
+1

**Comment by nmilojevic23** [comment]
+1

**Comment by JJEccles** [comment]
+1

**Comment by palash-fin** [comment]
++

**Comment by SMAntony** [comment]
+1

**Comment by celalettinbilgen2** [comment]
+1

**Comment by nerdalert** [comment]
+1

**Comment by akharlamov** [comment]
+1

**Comment by jacethemagician** [comment]
+2

**Comment by edwins** [comment]
+1

**Comment by esteban20304** [comment]
+1

**Comment by ynwang007** [comment]
> One work around that I notice vLLM already gives is using Docker containers. Use docker compose to serve multiple models that way two or more models can be served at a time based on your memory capabilities.

But does this approach support dynamic load/unload models based upon receiving the traffic?

**Comment by ynwang007** [comment]
+1

**Comment by MohamedAliRashad** [comment]
Everyone wants this feature and yet the issue is closed

**Comment by RealWorga** [comment]
> > One work around that I notice vLLM already gives is using Docker containers. Use docker compose to serve multiple models that way two or more models can be served at a time based on your memory capabilities.
> 
> But does this approach support dynamic load/unload models based upon receiving the traffic?

Can't you just monitor traffic and only keep up the container of the used models? If traffic points towards an inactive model, then you'd just have to suffer the loading at that timepoint

**Comment by prm-james-hill** [comment]
+1

**Comment by fti7** [comment]
+1

**Comment by lukas-kuhl** [comment]
+1

**Comment by chlebnymaciej** [comment]
+1

**Comment by abid-jawad** [comment]
+1

**Comment by mjtechguy** [comment]
+1 in 2025 🙏🏽

**Comment by nerdalert** [comment]
Throwing out the Podman incantation that I use in case anyone needs a jumpstart to run multiple instances on the same host via the container route. Pre-req is NVIDIA container toolkit for the containers to have access to the GPUs.

```bash
podman run --rm --device nvidia.com/gpu=1 \
    -e NVIDIA_VISIBLE_DEVICES=1 \
    -v /usr/local/cuda-12.4/lib64:/usr/local/cuda-12.4/lib64 \
    -v /var/home/cloud-user:/var/home/cloud-user \
    -p 8000:8000 \
    --ipc=host \
    vllm/vllm-openai:latest \
    --host 0.0.0.0 \
    --port 8000 \
    --model /PATH/TO/MODEL_SAFETENSORS \
    --load-format safetensors \
    --config-format hf \
    --trust-remote-code \
    --device cuda
    --served-model-name post-train
```

**Comment by FelixNeutatzMainWebSolutions** [comment]
+1

**Comment by OriginalByteMe** [comment]
+1

**Comment by igorperic17** [comment]
+1

**Comment by mkgs210** [comment]
+1

**Comment by SohrabTa** [comment]
+1

**Comment by 3rdAT** [comment]
+1

:)

**Comment by sticky-note** [comment]
+1

I am currently have a setup with:
- `liteLLM` serving as uniq endpoint
- `[llama-swap](https://github.com/mostlygeek/llama-swap)` profiles watching requests behind `liteLLM` and switches dynamic models if needed
- dynamics `vLLM` docker container initiated by `llama-swap` and served through `liteLLM` and `llama-swap` proxy.
- statics `vLLM` container initiated via `docker compose` and served through `liteLLM` directly.

It's a bit hacky but it currently works for my use case and really want to see this feature implemented in the best LLMs engine of the open-source market.
The `llama-swap` profiles logic is very limited that's why I ended with this setup which is still a little rigid
<3

**Comment by youkaichao** [comment]
the newest api server now has `/sleep` and `/wake_up` endpoint now, and you should be able to serve multiple models with the same gpu, and quickly switch the models.

**Comment by RealWorga** [comment]
> the newest api server now has `/sleep` and `/wake_up` endpoint now, and you should be able to serve multiple models with the same gpu, and quickly switch the models.

Please explain this further.

**Comment by abid-jawad** [comment]
> the newest api server now has `/sleep` and `/wake_up` endpoint now, and you should be able to serve multiple models with the same gpu, and quickly switch the models.

@youkaichao Does this imply that to switch models, I need to deactivate the current model and activate the next one, and I cannot run multiple models simultaneously in parallel?

**Comment by youkaichao** [comment]
> > the newest api server now has `/sleep` and `/wake_up` endpoint now, and you should be able to serve multiple models with the same gpu, and quickly switch the models.
> 
> [@youkaichao](https://github.com/youkaichao) Does this imply that to switch models, I need to deactivate the current model and activate the next one, and I cannot run multiple models simultaneously in parallel?

yes.

**Comment by MohamedAliRashad** [comment]
@youkaichao 
I run this command to vllm server `curl http://localhost:8000/sleep` and it gives me `{"detail":"Not Found"}`
I ran the vllm server with `--enable-sleep-mode` and without it and in both cases nothing changed

**Comment by youkaichao** [comment]
> [@youkaichao](https://github.com/youkaichao) I run this command to vllm server `curl http://localhost:8000/sleep` and it gives me `{"detail":"Not Found"}` I ran the vllm server with `--enable-sleep-mode` and without it and in both cases nothing changed

you also need the `VLLM_SERVER_DEV_MODE` environment variable.

**Comment by shohamyamin** [comment]
@youkaichao Is it possible that vllm will serve on the same container several models like ollama and the models will share the gpu and will be exposed in different ports?

**Comment by saad-ashfaq** [comment]
> [@youkaichao](https://github.com/youkaichao) I run this command to vllm server `curl http://localhost:8000/sleep` and it gives me `{"detail":"Not Found"}` I ran the vllm server with `--enable-sleep-mode` and without it and in both cases nothing changed

@MohamedAliRashad you need to use an HTTP POST request with the sleep level. 

Try `curl --data "level=1" http://localhost:8000/sleep` after running the vLLM server with `VLLM_SERVER_DEV_MODE=1` and `--enable-sleep-mode`.

**Comment by MohamedAliRashad** [comment]
@saad-ashfaq It works now, but if i sent a request after putting the model into sleep it gives me this error:
```
ERROR 03-25 07:49:03 [core.py:340] EngineCore hit an exception: Traceback (most recent call last):
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 333, in run_engine_core
ERROR 03-25 07:49:03 [core.py:340]     engine_core.run_busy_loop()
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 367, in run_busy_loop
ERROR 03-25 07:49:03 [core.py:340]     outputs = step_fn()
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 192, in step
ERROR 03-25 07:49:03 [core.py:340]     output = self.model_executor.execute_model(scheduler_output)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/executor/abstract.py", line 80, in execute_model
ERROR 03-25 07:49:03 [core.py:340]     output = self.collective_rpc("execute_model",
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/executor/uniproc_executor.py", line 56, in collective_rpc
ERROR 03-25 07:49:03 [core.py:340]     answer = run_method(self.driver_worker, method, args, kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/utils.py", line 2216, in run_method
ERROR 03-25 07:49:03 [core.py:340]     return func(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
ERROR 03-25 07:49:03 [core.py:340]     return func(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/worker/gpu_worker.py", line 242, in execute_model
ERROR 03-25 07:49:03 [core.py:340]     output = self.model_runner.execute_model(scheduler_output)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
ERROR 03-25 07:49:03 [core.py:340]     return func(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 1032, in execute_model
ERROR 03-25 07:49:03 [core.py:340]     hidden_states = self.model(
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 03-25 07:49:03 [core.py:340]     return self._call_impl(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 03-25 07:49:03 [core.py:340]     return forward_call(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/model_executor/models/llama.py", line 529, in forward
ERROR 03-25 07:49:03 [core.py:340]     model_output = self.model(input_ids, positions, intermediate_tensors,
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/compilation/decorators.py", line 245, in __call__
ERROR 03-25 07:49:03 [core.py:340]     model_output = self.forward(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/model_executor/models/llama.py", line 339, in forward
ERROR 03-25 07:49:03 [core.py:340]     def forward(
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 03-25 07:49:03 [core.py:340]     return self._call_impl(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 03-25 07:49:03 [core.py:340]     return forward_call(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 745, in _fn
ERROR 03-25 07:49:03 [core.py:340]     return fn(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 822, in call_wrapped
ERROR 03-25 07:49:03 [core.py:340]     return self._wrapped_call(self, *args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 400, in __call__
ERROR 03-25 07:49:03 [core.py:340]     raise e
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 387, in __call__
ERROR 03-25 07:49:03 [core.py:340]     return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 03-25 07:49:03 [core.py:340]     return self._call_impl(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 03-25 07:49:03 [core.py:340]     return forward_call(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "<eval_with_key>.66", line 208, in forward
ERROR 03-25 07:49:03 [core.py:340]     submod_1 = self.submod_1(getitem, s0, getitem_1, getitem_2, getitem_3);  getitem = getitem_1 = getitem_2 = submod_1 = None
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 822, in call_wrapped
ERROR 03-25 07:49:03 [core.py:340]     return self._wrapped_call(self, *args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 400, in __call__
ERROR 03-25 07:49:03 [core.py:340]     raise e
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 387, in __call__
ERROR 03-25 07:49:03 [core.py:340]     return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 03-25 07:49:03 [core.py:340]     return self._call_impl(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 03-25 07:49:03 [core.py:340]     return forward_call(*args, **kwargs)
ERROR 03-25 07:49:03 [core.py:340]   File "<eval_with_key>.2", line 5, in forward
ERROR 03-25 07:49:03 [core.py:340]     unified_attention_with_output = torch.ops.vllm.unified_attention_with_output(query_2, key_2, value, output_1, 'model.layers.0.self_attn.attn');  query_2 = key_2 = value = output_1 = unified_attention_with_output = None
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/_ops.py", line 1123, in __call__
ERROR 03-25 07:49:03 [core.py:340]     return self._op(*args, **(kwargs or {}))
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/attention/layer.py", line 363, in unified_attention_with_output
ERROR 03-25 07:49:03 [core.py:340]     self.impl.forward(self,
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/attention/backends/flash_attn.py", line 266, in forward
ERROR 03-25 07:49:03 [core.py:340]     flash_attn_varlen_func(
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/vllm/vllm_flash_attn/flash_attn_interface.py", line 173, in flash_attn_varlen_func
ERROR 03-25 07:49:03 [core.py:340]     out, softmax_lse = torch.ops._vllm_fa2_C.varlen_fwd(
ERROR 03-25 07:49:03 [core.py:340]   File "/usr/local/lib/python3.10/dist-packages/torch/_ops.py", line 1123, in __call__
ERROR 03-25 07:49:03 [core.py:340]     return self._op(*args, **(kwargs or {}))
ERROR 03-25 07:49:03 [core.py:340] RuntimeError: CUDA error: an illegal memory access was encountered
ERROR 03-25 07:49:03 [core.py:340] CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
ERROR 03-25 07:49:03 [core.py:340] For debugging consider passing CUDA_LAUNCH_BLOCKING=1
ERROR 03-25 07:49:03 [core.py:340] Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
ERROR 03-25 07:49:03 [core.py:340] 
ERROR 03-25 07:49:03 [core.py:340] 
CRITICAL 03-25 07:49:03 [core_client.py:269] Got fatal signal from worker processes, shutting down. See stack trace above for root cause issue.
Killed
```

It's not a big problem since the `/is_sleeping` endpoint is there but i still think the server should not crash when i send it a request while it's sleeping

**Comment by erdaltoprak** [comment]
In sleep mode with `--enable-sleep-mode` and `VLLM_SERVER_DEV_MODE=1` it reports
`vllm  | INFO 03-25 15:38:19 [worker.py:133] Sleep mode freed 20.54 GiB memory, 0.92 GiB memory is still in use`
Is there a reason for not totally freeing the memory ? @youkaichao

**Comment by youkaichao** [comment]
> Is there a reason for not totally freeing the memory ?

we only release kvcache and model weights. other parts of the memory are still alive.

**Comment by suparious** [comment]
I would beg to question why we have to provide the model name during inference, if vLLM only runs a single model. I have had to write all sorts of code to query the `/v1/models` endpoints to dynamically construct my inference requests without having to hardcode the name of the model everywhere.

**Comment by tjwebb** [comment]
+1

**Comment by rohan-uiuc** [comment]
Are there any plans to support serving multiple models on the same server? This seems like a must have for multi GPU clusters and is clearly a request from a lot of people.

**Comment by artist-genai** [comment]
> In sleep mode with `--enable-sleep-mode` and `VLLM_SERVER_DEV_MODE=1` it reports `vllm | INFO 03-25 15:38:19 [worker.py:133] Sleep mode freed 20.54 GiB memory, 0.92 GiB memory is still in use` Is there a reason for not totally freeing the memory ? [@youkaichao](https://github.com/youkaichao)
> 
> I also created a bug issue for the crash on request if it's sleeping, I'm willing to contribute and create the necessary PR if the impacted endpoints are clearly defined ([#15483](https://github.com/vllm-project/vllm/issues/15483))

Were you able to run this in docker? When I am trying to do so, all I am getting is that it is an unknown flag.
```
docker run -d --name vllmeta --runtime=nvidia --gpus all \
    -v ~/.cache/huggingface:/root/.cache/huggingface \
    --env "HUGGING_FACE_HUB_TOKEN=<Token>" \
    --env "VLLM_SERVER_DEV_MODE=1" \
    --enable-sleep-mode \
    -p 8000:8000 \
    --ipc=host \
    vllm/vllm-openai:latest \
    --model meta-llama/Llama-3.1-8B-Instruct\
    --gpu-memory-utilization 0.5 \
    --quantization bitsandbytes \
    --dtype float16 \
    --enforce-eager \
    --max-model-len 2048
unknown flag: --enable-sleep-mode

Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]

Run 'docker run --help' for more information
```

**Comment by Snify89** [comment]
> > In sleep mode with `--enable-sleep-mode` and `VLLM_SERVER_DEV_MODE=1` it reports `vllm | INFO 03-25 15:38:19 [worker.py:133] Sleep mode freed 20.54 GiB memory, 0.92 GiB memory is still in use` Is there a reason for not totally freeing the memory ? [@youkaichao](https://github.com/youkaichao)
> > I also created a bug issue for the crash on request if it's sleeping, I'm willing to contribute and create the necessary PR if the impacted endpoints are clearly defined ([#15483](https://github.com/vllm-project/vllm/issues/15483))
> 
> Were you able to run this in docker? When I am trying to do so, all I am getting is that it is an unknown flag.
> 
> ```
> docker run -d --name vllmeta --runtime=nvidia --gpus all \
>     -v ~/.cache/huggingface:/root/.cache/huggingface \
>     --env "HUGGING_FACE_HUB_TOKEN=<Token>" \
>     --env "VLLM_SERVER_DEV_MODE=1" \
>     --enable-sleep-mode \
>     -p 8000:8000 \
>     --ipc=host \
>     vllm/vllm-openai:latest \
>     --model meta-llama/Llama-3.1-8B-Instruct\
>     --gpu-memory-utilization 0.5 \
>     --quantization bitsandbytes \
>     --dtype float16 \
>     --enforce-eager \
>     --max-model-len 2048
> unknown flag: --enable-sleep-mode
> 
> Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]
> 
> Run 'docker run --help' for more information
> ```

--enable-sleep-mode is a parameter but you have put it in docker run parameters.


 docker run -d --name vllmeta --runtime=nvidia --gpus all \
     -v ~/.cache/huggingface:/root/.cache/huggingface \
     --env "HUGGING_FACE_HUB_TOKEN=<Token>" \
     --env "VLLM_SERVER_DEV_MODE=1" \
     -p 8000:8000 \
     --ipc=host \
     vllm/vllm-openai:latest \
     --model meta-llama/Llama-3.1-8B-Instruct\
     --gpu-memory-utilization 0.5 \
     --quantization bitsandbytes \
     --dtype float16 \
     --enforce-eager \
     --max-model-len 2048 \
     --enable-sleep-mode

**Comment by felix0080** [comment]
+1

**Comment by huakyouin** [comment]
+1

**Comment by b-d-e** [comment]
+1

**Comment by thedigitalme** [comment]
+1

**Comment by 2abet** [comment]
I wonder if this can be done in the vllm production stack

**Comment by kfletche** [comment]
Love to see this happen.

**Comment by Hasen-dc** [comment]
I want to run two models on a GPU server using vLLM. The larger model is used for dialogue, while the smaller one is used for intent recognition, sentiment analysis, etc.

**Comment by chengchengpei** [comment]
+1

i need this feature too. I am working on this now :)

**Comment by Chen-zexi** [comment]
I am planning to support multiple model serving in [vLLM-CLI](https://github.com/Chen-zexi/vllm-cli). Feel free to share any additional feature request so I can better implement it!

**Comment by JonasWild** [comment]
+1

**Comment by saeid93** [comment]
+1

**Comment by jiarong0907** [comment]
Hi folks who need multi-model serving using vllm, our kvcached project can support that by installing a plugin into vLLM. 

https://github.com/ovg-project/kvcached.

**Comment by gboksm11** [comment]
+2 (me and aris)

**Comment by bogan-FMA** [comment]
or this can be done by using GPU virtualization such as orionx to run multiple vllm instance on one GPU.

**Comment by Absoler** [comment]
I saw the sleep/wakeup could do the similar things from [https://blog.vllm.ai/2025/10/26/sleep-mode.html](url).
